quixstreams/sources/manager.py

```diff
@@ -0,0 +1,255 @@
+        configure_logging(self._loglevel, str(self.source), pid=True)
+        logger.info("Source started")
+
+        try:
+            try:
+                self.source.run()
+            except BaseException:
+                self._cleanup(failed=True)
+                raise
+            else:
+                self._cleanup(failed=False)
+        except BaseException as err:
+            logger.exception(f"Error in source")
+            self._report_exception(err)
+
+        logger.info("Source completed")
+        threadcount = threading.active_count()
+        logger.debug(
+            "%s active thread%s in source",
+            threadcount,
+            "s" if threadcount > 1 else "",
+        )
+
+    def _cleanup(self, failed: bool) -> None:
+        """
+        Execute post-run cleanup
+        """
+        logger.debug("Cleaning up source")
+        self.source.cleanup(failed)
+
+    def _stop(self, signum, _):
+        """
+        Ask the source to stop execution
+        """
+        signame = signal.Signals(signum).name
+        logger.debug("Source received %s, stopping", signame)
+        if self._stopping:
+            return
+
+        self._stopping = True
+
+        try:
+            self.source.stop()
+        except BaseException as err:
+            logger.exception("Error stopping source")
+            self._report_exception(err)
+
+    def _report_exception(self, err: Exception) -> None:
+        """
+        Write exception to the pipe so the parent process can access it.
+        If the exception can't be pickled (for example confluent-kafka C exceptions) raise a RuntimeError.
+        """
+        try:
+            self._wpipe.send(err)
+        except (PicklingError, TypeError):
+            self._wpipe.send(RuntimeError(str(err)))
+
+    # --- PARENT PROCESS METHODS --- #
+
+    def start(self) -> "SourceProcess":
+        logger.info("Starting source %s", self.source)
+        return super().start()
+
+    def raise_for_error(self) -> bool:
+        """
+        If the child process has terminated with an exception raises a
+        :class:`quixstreams.sources.manager.SourceException`.
+        """
+        if super().is_alive():
+            return
+
+        if not self._exceptions:
+            while self._rpipe.poll():
+                try:
+                    self._exceptions.append(self._rpipe.recv())
+                except EOFError:
+                    return
+
+        if self._exceptions:
+            raise SourceException(self) from self._exceptions[-1]
+
+        if self.exitcode not in (0, -signal.SIGTERM):
+            raise SourceException(self)
+
+    def stop(self):
+        """
+        Handle shutdown of the source and the associated process
+
+        First try a graceful shutdown by sending a SIGTERM and waiting up to
+        `source.shutdown_timeout` seconds for the process to exit. If the process
+        is still alive force kill it with a SIGKILL/
```

Comment:
```suggestion
        Handle shutdown of the source and its subprocess.

        First, it tries to shut down gracefully by sending a SIGTERM and waiting up to
        `source.shutdown_timeout` seconds for the process to exit. If the process
        is still alive, it will kill it with a SIGKILL.
```